src/host/_stream.cpp

```diff
@@ -126,565 +165,286 @@ void AdjustCursorPosition(SCREEN_INFORMATION& screenInfo, _In_ til::point coordC
+                        const auto previousColumn = pos.x;
+                        pos.x = textBuffer.GetRowByOffset(pos.y).NavigateToPrevious(previousColumn);
+
+                        RowWriteState state{
+                            .text = { &tabSpaces[0], 8 },
+                            .columnBegin = pos.x,
+                            .columnLimit = previousColumn,
+                        };
+                        textBuffer.Write(pos.y, textBuffer.GetCurrentAttributes(), state);
+                        numSpaces -= previousColumn - pos.x;
                     }
-                }
-                else if (IS_CONTROL_CHAR(LastChar))
-                {
-                    CursorPosition.x -= 1;
-                    TempNumSpaces -= 1;
 
-                    // overwrite second character of ^x sequence.
-                    if (dwFlags & WC_DESTRUCTIVE_BACKSPACE)
+                    // The cursor movement logic is a little different for the last iteration, so we exit early here.
+                    glyphCount--;
+                    if (glyphCount <= 0)
                     {
-                        try
-                        {
-                            screenInfo.Write(OutputCellIterator(UNICODE_SPACE, Attributes, 1), CursorPosition);
-                            Status = STATUS_SUCCESS;
-                        }
-                        CATCH_LOG();
+                        break;
                     }
 
-                    CursorPosition.x -= 1;
-                }
-                else if (IsGlyphFullWidth(LastChar))
-                {
-                    CursorPosition.x -= 1;
-                    TempNumSpaces -= 1;
-
-                    AdjustCursorPosition(screenInfo, CursorPosition, dwFlags & WC_KEEP_CURSOR_VISIBLE, psScrollY);
-                    if (dwFlags & WC_DESTRUCTIVE_BACKSPACE)
+                    // Otherwise, in case we need to delete 2 or more glyphs, we need to ensure we properly wrap lines back up.
+                    if (pos.x == 0 && pos.y != 0)
                     {
-                        try
-                        {
-                            screenInfo.Write(OutputCellIterator(UNICODE_SPACE, Attributes, 1), CursorPosition);
-                            Status = STATUS_SUCCESS;
-                        }
-                        CATCH_LOG();
+                        moveUp();
                     }
-                    CursorPosition.x -= 1;
+                }
+
+                // After the last iteration the cursor might now be in the first column after a line
+                // that was previously padded with a whitespace in the last column due to a wide glyph.
+                // Now that the wide glyph is presumably gone, we can move up a line.
+                if (pos.x == 0 && pos.y != 0 && textBuffer.GetRowByOffset(pos.y - 1).WasDoubleBytePadded())
+                {
+                    moveUp();

```

Comment:
I thought moveUp handled this, but I am guessing that we don't want to *always* move up just only sometimes

Reply:
Yeah in this case we want to unwrap the line (move up), only if the previous line was wide-glyph padded. The idea is that we get to this line _after_ already removing a glyph. In this case we might have moved from a `pos.x` of 2 to a `pos.x` of 0, like deleting the first glyph of a line. Normally we wouldn't unwrap the line now, but if the line is only wrapped because it was wide-glyph padded, then the previous padding whitespace doesn't actually exist. 😅